cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesService.java

```diff
@@ -90,60 +95,85 @@ public PreferencesService(DatasetFramework datasetFramework, TransactionSystemCl
-        adminEventPublisher.publishProfileAssignment(entityId, profile.get());
+      if (cause instanceof ProfileConflictException) {
+        throw (ProfileConflictException) cause;
       }
+      if (cause instanceof BadRequestException) {
+        throw (BadRequestException) cause;
+      }
+    }
+  }
 
-      // if old properties has the profile, remove the association
-      if (oldProfile.isPresent()) {
-        profileDataset.removeProfileAssignment(oldProfile.get(), entityId);
+  /**
+   * Validate the profile status is enabled and set the preferences in same transaction
+   */
+  private void setConfig(ProfileDataset profileDataset, PreferencesDataset preferencesDataset, EntityId entityId,
+                         Map<String, String> propertyMap)
+    throws NotFoundException, ProfileConflictException, BadRequestException {
+
+    boolean isInstanceLevel = entityId.getEntityType().equals(EntityType.INSTANCE);
+    NamespaceId namespaceId = isInstanceLevel ?
+      NamespaceId.SYSTEM : ((NamespacedEntityId) entityId).getNamespaceId();
+
+    // validate the profile and publish the necessary metadata change if the profile exists in the property
+    Optional<ProfileId> profile = SystemArguments.getProfileIdFromArgs(namespaceId, propertyMap);
+    if (profile.isPresent()) {
+      ProfileId profileId = profile.get();
+      // if it is instance level set, the profile has to be in SYSTEM scope, so throw BadRequestException when
+      // setting a USER scoped profile
+      if (isInstanceLevel && !propertyMap.get(SystemArguments.PROFILE_NAME).startsWith(EntityScope.SYSTEM.name())) {
+        throw new BadRequestException(String.format("Cannot set profile %s at the instance level. " +
+                                                      "Only system profiles can be set at the instance level. " +
+                                                      "The profile property must look like SYSTEM:[profile-name]",
+                                                    propertyMap.get(SystemArguments.PROFILE_NAME)));
       }
 
-      // if new profiles do not have profile information but old profiles have, it is same as deletion of the profile
-      if (!profile.isPresent() && oldProfile.isPresent()) {
-        adminEventPublisher.publishProfileUnAssignment(entityId);
+      if (profileDataset.getProfile(profileId).getStatus() == ProfileStatus.DISABLED) {
+        throw new ProfileConflictException(String.format("Profile %s in namespace %s is disabled. It cannot be " +
+                                                           "assigned to any programs or schedules",
+                                                         profileId.getProfile(), profileId.getNamespace()),
+                                           profileId);
       }
-    }, NotFoundException.class, ProfileConflictException.class);
+
+    }
+
+    // need to get old property and check if it contains profile information
```

Comment:
where is the check done ? is the check done when we check `oldProfile.isPresent` ?

Reply:
yes